tinygrad/shape/symbolic.py

```diff
@@ -10,6 +10,8 @@
 # NOTE: Python has different behavior for negative mod and floor div than c
 # symbolic matches the Python behavior, but the code output is agnostic, and will never have negative numbers in div or mod
 
+infer_cache: Dict[Tuple[int, Any], int] = {}
```

Comment:
Does this have to be global?

Reply:
In this impl it has to be. The problem is that sym_infer is really expensive (takes up to 50% of the gpt2 step) we can cache results, since they are reused often. gpt2 runs several TinyJits, so wanna use the same cache across them. So I see only 2 solutions global var (maybe move to jit.py) or `JitCtx` which is also shared between TinyJits.

As for JitCtx impl: I have chosen JitCtx since it's the most opaque for the end user as they should not care about caches. `jit_ctx = JitCtx({pos: start_pos})` creates a shadow cache for the provided set of vars which is shared between TinyJits (and when you want a new set of values to pass you just create a new `JitCtx` so caches do not intersect).

Reply:
With #1896 i think we can drop and have cache just inside TinyJit. I think the approach is possible to apply to llama as well to have one Jit.

Reply:
Looking at #1896 I think of the following plan. This PR uses one JIT for the whole gpt step which allows to solve problem with caching having no new globals and no JitCtx. But the main problem is that it kinda delay execution of kernels on gpus and we haven't got cpu+gpu working in parallel (this is really visible with hipgraphs). The solution for this is a small refactor of JIT to allow updating batches of kernels and launch them immediately (now we just update everything and launch them after that).

I am not sure it's a great idea to mix up refactor of JIT with this PR, since this PR is already big. So we might have not the best performance for now and improve it with the JIT refactor. How that sounds?

Reply:
Put it into JIT itself